File: jpi2/src/main/java/org/jenkinsci/gradle/plugins/jpi2/accmod/CheckAccessModifierTask.java
package org.jenkinsci.gradle.plugins.jpi2.accmod;

import org.gradle.api.DefaultTask;
import org.gradle.api.file.ConfigurableFileCollection;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.provider.MapProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Classpath;
import org.gradle.api.tasks.CompileClasspath;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.OutputDirectory;
import org.gradle.api.tasks.TaskAction;
import org.gradle.workers.WorkerExecutor;

import javax.inject.Inject;

public abstract class CheckAccessModifierTask extends DefaultTask {
    public static final String NAME = "checkAccessModifier";
    public static final String PREFIX = NAME + ".";

    private final WorkerExecutor workerExecutor;
    private final ConfigurableFileCollection accessModifierClasspath;
    private final MapProperty<String, Object> accessModifierProperties;
    private final ConfigurableFileCollection compileClasspath;
    private final ConfigurableFileCollection compilationDirs;
    private final Property<Boolean> ignoreFailures;
    private final DirectoryProperty outputDirectory;

    @Inject
    public CheckAccessModifierTask(WorkerExecutor workerExecutor) {
        this.workerExecutor = workerExecutor;
        var objects = getProject().getObjects();
        this.accessModifierClasspath = objects.fileCollection();
        this.accessModifierProperties = objects.mapProperty(String.class, Object.class);
        this.compileClasspath = objects.fileCollection();
        this.compilationDirs = objects.fileCollection();
        this.ignoreFailures = objects.property(Boolean.class);
        this.outputDirectory = objects.directoryProperty();
    }

    @Classpath
    public ConfigurableFileCollection getAccessModifierClasspath() {
        return accessModifierClasspath;
    }

    @Input
    public MapProperty<String, Object> getAccessModifierProperties() {
        return accessModifierProperties;
    }

    @CompileClasspath
    public ConfigurableFileCollection getCompileClasspath() {
        return compileClasspath;
    }

    @InputFiles
    public ConfigurableFileCollection getCompilationDirs() {
        return compilationDirs;
    }

    @Input
    public Property<Boolean> getIgnoreFailures() {
        return ignoreFailures;
    }

    @OutputDirectory
    public DirectoryProperty getOutputDirectory() {
        return outputDirectory;
    }

    @TaskAction
    public void check() {
        var queue = workerExecutor.classLoaderIsolation(spec -> spec.getClasspath().from(accessModifierClasspath));
        for (var compilationDir : compilationDirs) {
            String parentName = compilationDir.getParentFile() == null ? "classes" : compilationDir.getParentFile().getName();
            String fileName = compilationDir.getName() + "-" + parentName + ".txt";
            queue.submit(CheckAccess.class, params -> {
                params.getClasspathToScan().from(compilationDirs, compileClasspath);
                params.getDirToCheck().set(compilationDir);
                params.getIgnoreFailures().set(ignoreFailures);
                params.getPropertiesForAccessModifier().set(accessModifierProperties);
                params.getOutputFile().set(outputDirectory.file(fileName));
            });
        }
    }
}

File: src/main/kotlin/org/jenkinsci/gradle/plugins/accmod/CheckAccessModifierTask.kt
package org.jenkinsci.gradle.plugins.accmod

import org.gradle.api.DefaultTask
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.plugins.JavaPluginExtension
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Classpath
import org.gradle.api.tasks.CompileClasspath
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.Nested
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.TaskAction
import org.gradle.jvm.toolchain.JavaLanguageVersion
import org.gradle.jvm.toolchain.JavaLauncher
import org.gradle.jvm.toolchain.JavaToolchainService
import org.gradle.jvm.toolchain.JvmVendorSpec
import org.gradle.kotlin.dsl.getByType
import org.gradle.kotlin.dsl.submit
import org.gradle.workers.WorkerExecutor
import javax.inject.Inject

/**
 * This task is modeled on org.kohsuke:access-modifier-checker
 *
 * @see org.kohsuke.accmod.impl.EnforcerMojo
 */
abstract class CheckAccessModifierTask @Inject constructor(
        private val workerExecutor: WorkerExecutor,
        javaToolchainService: JavaToolchainService,
) : DefaultTask() {
    companion object {
        const val NAME = "checkAccessModifier"
        const val PREFIX = "$NAME."
    }

    @get:Classpath
    abstract val accessModifierClasspath: ConfigurableFileCollection

    @get:Input
    abstract val accessModifierProperties: MapProperty<String, Any>

    @get:CompileClasspath
    abstract val compileClasspath: ConfigurableFileCollection

    @get:InputFiles
    abstract val compilationDirs: ConfigurableFileCollection

    @get:Input
    abstract val ignoreFailures: Property<Boolean>

    @get:OutputDirectory
    abstract val outputDirectory: DirectoryProperty

    @get:Nested
    abstract val launcher: Property<JavaLauncher>

    init {
        val toolchain = project.extensions.getByType<JavaPluginExtension>().toolchain
        val requestedVersion = toolchain.languageVersion.map { it.asInt() }.orNull
        val defaultLauncher = if (requestedVersion == null || requestedVersion < 11) {
            javaToolchainService.launcherFor {
                languageVersion.set(JavaLanguageVersion.of(11))
            }
        } else {
            javaToolchainService.launcherFor(toolchain)
        }
        launcher.convention(defaultLauncher)
    }

    @TaskAction
    fun check() {
        val q = workerExecutor.processIsolation {
            forkOptions {
                this.executable = launcher.map { it.executablePath.toString() }.get()
            }
            classpath.from(accessModifierClasspath)
        }
        for (compilationDir in compilationDirs) {
            q.submit(CheckAccess::class) {
                classpathToScan.from(compilationDirs, compileClasspath)
                dirToCheck.set(compilationDir)
                ignoreFailures.set(this@CheckAccessModifierTask.ignoreFailures)
                propertiesForAccessModifier.set(accessModifierProperties)
                outputFile.set(outputDirectory.file("${compilationDir.name}-${compilationDir.parentFile.name}.txt"))
            }
        }
    }
}
